rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md

```diff
@@ -0,0 +1,542 @@
+  callback functions for checkpointing and TensorBoard setup.
+
+* The return value is flexible. It can be a single value, or a list of values,
+  or a dictionary of metrics, or even a `History` object returned by
+  `model.fit()`.
+
+* The `hypermodel` and `objective` can be optional. The user doesn't need to
+  define a `HyperModel`. If the return value is a single value, it will be
+  minimized by default. Therefore, objective is also optional.
+
+* The user can build a unique path to save each model with `trial.trial_id`.
+
+For the use case of reusing existing Keras code. The user can use the following
+workflow, which calls a function using all the hyperparameters. The user only
+needs to write a function to call the existing Keras code and return the
+evaluation results.
+
+```py
+class MyTuner(kt.Tuner):
+ def run_trial(self, trial, callbacks=None, dataset=None, **kwargs):
+   hp = trial.hyperparameters
+   return build_and_evaluate_model(
+       hp.Int("num_features", 10, 15),
+       hp.Choice('units', [8, 16, 32]),
+       ...
+       trial.trial_id,
+   ))
+   # Save model can be handled by the user.
+   # `trial_id` is unique for each trial.
+
+tuner = MyTuner(...)
+tuner.search()
+# Retraining the model
+build_and_evaluate_model(**tuner.get_best_hyperparameters()[0])
+```
+	
+
+In this workflow, the user can easily retrain the model by calling the function again with the best hyperparameters.
+
+### The HyperModel workflow
+
+For users who prefer to follow the old workflow, they can also implement the HyperModel above by overriding the build function and the fit function. The build function builds and returns the model. The fit function does the data preprocessing and model training.
+
+Following is a code example implementing the same functionality of the code example above.
+
+```py
+import numpy as np
+import keras_tuner as kt
+from tensorflow import keras
+
+class MyHyperModel(kt.HyperModel):
+
+  def build(self, hp):
+    # Model building
+    # Input shape depends on a hyperparameter used by data preprocessing.
+    inputs = keras.Input(shape=(hp.Int("num_features", 10, 15),))
+    x = keras.layers.Dense(
+        hp.Choice('units', [8, 16, 32]),
+        activation='relu')(inputs)
+    outputs = keras.layers.Dense(1, activation='relu')(x)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    model.compile(loss='mse',
+                  metrics=['mae'])
+    return model
+  
+  def fit(self, hp, model, dataset=None, callbacks=None, **kwargs):
+    # Data preprocessing
+    # Get the hyperparameter value used in `build()`.
+    dataset = dataset.map(hp.get("num_features"))
+    # Model training
+    # Returning the training history
+    # or a similar dictionary if using custom training loop.
+    return model.fit(
+        dataset,
+        epochs=100,
+        validation_data=validation_data,
+        # Tune whether to use shuffle.
+        shuffle=hp.Boolean("shuffle"),
```

Comment:
If it's a dataset, you can't use this arg. Shuffling happens on the dataset object.

Reply:
Changed to not using datasets to show case tuning fit args.